File: csharp/cache/LFUReplacementPolicy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace CacheSystem
{
    /// <summary>
    /// Least Frequently Used (LFU) cache replacement policy.
    /// </summary>
    public class LFUReplacementPolicy<KeyType> : IReplacementPolicy<KeyType>
    {
        private readonly Dictionary<KeyType, int> _keyFrequency = new Dictionary<KeyType, int>();
        private readonly SortedDictionary<int, HashSet<KeyType>> _frequencyKeys = new SortedDictionary<int, HashSet<KeyType>>();
        private readonly object _lock = new object();
        private int _minFrequency = 1;

        public bool Add(KeyType key)
        {
            lock (_lock)
            {
                if (_keyFrequency.ContainsKey(key))
                {
                    return false;
                }
                _keyFrequency[key] = 1;
                if (!_frequencyKeys.ContainsKey(1))
                {
                    _frequencyKeys[1] = new HashSet<KeyType>();
                }
                _frequencyKeys[1].Add(key);
                _minFrequency = 1;
                return true;
            }
        }

        public bool Remove(KeyType key)
        {
            lock (_lock)
            {
                if (!_keyFrequency.TryGetValue(key, out var freq))
                {
                    return false;
                }
                _keyFrequency.Remove(key);
                _frequencyKeys[freq].Remove(key);
                if (_frequencyKeys[freq].Count == 0)
                {
                    _frequencyKeys.Remove(freq);
                    if (_minFrequency == freq)
                    {
                        _minFrequency = _frequencyKeys.Keys.FirstOrDefault();
                    }
                }
                return true;
            }
        }

        public bool Access(KeyType key)
        {
            lock (_lock)
            {
                if (!_keyFrequency.TryGetValue(key, out var freq))
                {
                    return false;
                }
                _keyFrequency[key] = freq + 1;
                _frequencyKeys[freq].Remove(key);
                if (_frequencyKeys[freq].Count == 0)
                {
                    _frequencyKeys.Remove(freq);
                    if (_minFrequency == freq)
                    {
                        _minFrequency++;
                    }
                }
                if (!_frequencyKeys.ContainsKey(freq + 1))
                {
                    _frequencyKeys[freq + 1] = new HashSet<KeyType>();
                }
                _frequencyKeys[freq + 1].Add(key);
                return true;
            }
        }

        public async Task<KeyType> EvictAsync()
        {
            lock (_lock)
            {
                if (_frequencyKeys.Count == 0)
                {
                    throw new CacheException("No keys to evict.");
                }
                var keySet = _frequencyKeys[_minFrequency];
                var key = keySet.First();
                keySet.Remove(key);
                if (keySet.Count == 0)
                {
                    _frequencyKeys.Remove(_minFrequency);
                }
                _keyFrequency.Remove(key);
                return key;
            }
        }
    }
}

File: python/cache/lfu_replacement_policy.py
# lfu_replacement_policy.py
from collections import defaultdict
from threading import RLock
from typing import Generic, TypeVar, Optional

from replacement_policy import IReplacementPolicy

KeyType = TypeVar("KeyType")


class LFUReplacementPolicy(IReplacementPolicy[KeyType], Generic[KeyType]):
    """Least Frequently Used (LFU) cache replacement policy."""

    def __init__(self):
        self._lock = RLock()
        self._key_freq = defaultdict(int)
        self._freq_keys = defaultdict(set)
        self._min_freq = 0

    def add(self, key: KeyType) -> bool:
        with self._lock:
            if key in self._key_freq:
                return False
            self._key_freq[key] = 1
            self._freq_keys[1].add(key)
            self._min_freq = 1
            return True

    def remove(self, key: KeyType) -> bool:
        with self._lock:
            if key not in self._key_freq:
                return False
            freq = self._key_freq.pop(key)
            self._freq_keys[freq].remove(key)
            if not self._freq_keys[freq]:
                del self._freq_keys[freq]
                if freq == self._min_freq:
                    self._min_freq += 1
            return True

    def access(self, key: KeyType) -> bool:
        with self._lock:
            if key not in self._key_freq:
                return False
            freq = self._key_freq[key]
            self._key_freq[key] += 1
            self._freq_keys[freq].remove(key)
            if not self._freq_keys[freq]:
                del self._freq_keys[freq]
                if freq == self._min_freq:
                    self._min_freq += 1
            self._freq_keys[freq + 1].add(key)
            return True

    def evict(self) -> Optional[KeyType]:
        with self._lock:
            if not self._key_freq:
                return None
            key = next(iter(self._freq_keys[self._min_freq]))
            self.remove(key)
            return key
